### PR TITLE
DEV-4426 - Klarna HPP session & payment checker

### PR DESCRIPTION
**Pull request by alexbowers**

### CODE REVIEW

File: src/Message/KPRefundOrderCheckResponse.php [hunk 1]
```diff
@@ -31,6 +31,11 @@ public function getData(): array
     public function isSuccessful()
     {
         $klarnaOrder = $this->responseBody;
+
+        if ($this->response->getStatusCode() != 200) {
+            return $this->response->getStatusCode();
+        }
+
```
Comment (ProjectZero4) [inline src/Message/KPRefundOrderCheckResponse.php]: Looks like this function is meant to return a bool, this is now returning a bool or integer?

**Review by ProjectZero4 — CHANGES_REQUESTED** [review]

File: src/Message/KPRefundOrderCheckResponse.php (hunk 1 above)
Comment (ProjectZero4) [reply src/Message/KPRefundOrderCheckResponse.php]: Can we not do this a better way?

File: src/Message/KPRefundOrderCheckResponse.php [hunk 2]
```diff
@@ -31,6 +31,11 @@ public function getData(): array
     public function isSuccessful()
     {
         $klarnaOrder = $this->responseBody;
+
+        if ($this->response->getStatusCode() != 200) {
+            return $this->response->getStatusCode();
```
Comment (alexbowers) [inline src/Message/KPRefundOrderCheckResponse.php]:
```suggestion
            return false;
```

File: src/Message/KPRefundOrderCheckResponse.php (hunk 1 above)
Comment (alexbowers) [reply src/Message/KPRefundOrderCheckResponse.php]: Good spot, this can just be `return false`

**Comment by JNSB22** [comment]
👍

**Review by Marcus-Nightingale — APPROVED** [review]

File: src/Message/KPRefundOrderCheckResponse.php [hunk 3]
```diff
@@ -29,12 +30,33 @@ public function getData(): array
...
 
-        if ($this->response->getStatusCode() != 200) {
-            return $this->response->getStatusCode();
+        $refundedAmount = 0;
+        if (!empty($klarnaOrder) && !empty($klarnaOrder['refunds'])) {
+            foreach ($klarnaOrder['refunds'] as $refund) {
+                $refundedAmount += $refund['refunded_amount'];
+            }
         }
+        $remainingAmount = $klarnaOrder['captured_amount'] - $refundedAmount;
```
Comment (ProjectZero4) [inline src/Message/KPRefundOrderCheckResponse.php]: Line 47 is not wrapped in the `!empty()` check. Is there a possiblity it could be empty here?

**Review by ProjectZero4 — CHANGES_REQUESTED** [review]

File: src/Message/KPRefundOrderCheckResponse.php (hunk 3 above)
Comment (alexbowers) [reply src/Message/KPRefundOrderCheckResponse.php]: So, I think as long as you get a successful response back from Klarna, these properties are always defined. I tried to minimise the amount of changes made; and instead kept it as close to the existing code as possible. I am happy to change this if necessary, but I think it's fine

File: src/Message/HPPSessionCheckResponse.php [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+use Omnipay\Common\Message\RequestInterface;
+use Omnipay\Common\Message\ResponseInterface;
+
+class HPPSessionCheckResponse extends AbstractResponse implements ResponseInterface
+{
+    protected $request;
+    protected $response;
+    protected $responseBody;
+
+    public function __construct(RequestInterface $request, $response, $refundAmount)
```
Comment (jamie-peck) [inline src/Message/HPPSessionCheckResponse.php]: How come we dont use $refundAmount here?

File: src/Message/KPRefundOrderCheckResponse.php [hunk 5]
```diff
@@ -29,8 +30,34 @@ public function getData(): array
...
     public function isSuccessful()
+    {
+        return $this->response->getStatusCode() == 200;
+    }
+
+    public function isRefundAllowed()
+    {
+        $klarnaOrder = $this->responseBody;
+
+        $refundedAmount = 0;
```
Comment (jamie-peck) [inline src/Message/KPRefundOrderCheckResponse.php:41]: Is `$this->refundAmount` not accurate here? Seems like we are re-calculating it?

Comment (alexbowers) [reply src/Message/KPRefundOrderCheckResponse.php:41]: `$this->refundAmount` is the amount the customer _wants_ to refund, we're calculating how much they're allowed to refund (ie, if a partial refund has taken place before)

**Review by ProjectZero4 — APPROVED** [review]

**Pull request merged**